### PR TITLE
[CY-10243] asciicast support for guac

### PR DESCRIPTION
**Pull request by amityahav**
- Added support for asciicast recording in guacd's SSH proxy. the same logic that is implemented in the native SSH is also applied here.

Tested locally

### CODE REVIEW

File: src/libguac/guacamole/recording.h [hunk 1]
```diff
@@ -97,6 +97,37 @@ typedef struct guac_recording {
 
 } guac_recording;
 
+
```
Comment (amityahav) [inline src/libguac/guacamole/recording.h:100]: just moved

File: src/libguac/timestamp.c [hunk 2]
```diff
@@ -68,5 +68,4 @@ void guac_timestamp_msleep(int duration) {
     /* Sleep for specified interval */
     nanosleep(&sleep_period, NULL);
 
-}
-
+}
```
Comment (Virviil) [inline src/libguac/timestamp.c:71]:
```suggestion
}

```

File: src/common-ssh/asciicast.c [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    cJSON_AddItemToObject(env, "TERM", term);
+    cJSON_AddItemToObject(header, "version", version);
+    cJSON_AddItemToObject(header, "width", width);
+    cJSON_AddItemToObject(header, "height", height);
+    cJSON_AddItemToObject(header, "timestamp", timestamp);
+    cJSON_AddItemToObject(header, "env", env);
+
+    json = cJSON_PrintUnformatted(header);
+
+    end:
```
Comment (Virviil) [inline src/common-ssh/asciicast.c:263]: Please check free sequeences here also

File: src/common-ssh/asciicast.c [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        cJSON_Delete(header);
+        return json;
+}
```
Comment (Virviil) [inline src/common-ssh/asciicast.c:266]:
```suggestion
}

```

File: src/common-ssh/common-ssh/asciicast.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+asciicast_recording* asciicast_recording_create(char* path, char* name, int create_path, guac_client *client);
+char* create_asciicast_header(asciicast_recording *rec);
+char save_asciicast_file(asciicast_recording *rec, guac_client *client);
+void free_asciicast_recording(asciicast_recording *rec);
+
+
+char write_cast_event(float sec, const char* mode, const char* buffer, int bytes_read, guac_client* client, asciicast_recording* rec);
+char *asciicast_event_to_json(float timestamp, const char* mode, const char* data);
+
+#endif
```
Comment (Virviil) [inline src/common-ssh/common-ssh/asciicast.h:34]:
```suggestion
#endif

```

File: src/protocols/ssh/ssh.c [hunk 6]
```diff
@@ -543,8 +646,14 @@ void* ssh_client_thread(void* data) {
...
+            guac_client_log(client, GUAC_LOG_ERROR, 
+            "failed to create asciicast file for %s", ssh_client->ascii_recording->name);
+        }
+    }
 
-}
+    guac_client_log(client, GUAC_LOG_INFO, "SSH connection ended.");
 
+    return NULL;
+}
```
Comment (Virviil) [inline src/protocols/ssh/ssh.c:659]:
```suggestion
}

```

**Review by Virviil — CHANGES_REQUESTED** [review]

**Review by Virviil — APPROVED** [review]

**Pull request merged**